ql/src/java/org/apache/hadoop/hive/ql/parse/DeleteSemanticAnalyzer.java

```diff
@@ -0,0 +1,165 @@
+      return;
+    }
+
+    Rewriter<DeleteStatement> rewriter = rewriterFactory.createRewriter(table, getFullTableNameForSQL(tableName));
+    ParseUtils.ReparseResult rr = rewriter.rewrite(ctx, new DeleteStatement(table, where));
+
+    Context rewrittenCtx = rr.rewrittenCtx;
+    ASTNode rewrittenTree = rr.rewrittenTree;
+
+    analyzeRewrittenTree(rewrittenTree, rewrittenCtx);
+
+    updateOutputs(table);
+  }
+
+  private void genTruncatePlan(Table table, ASTNode tabNameNode) throws SemanticException {
+    String rewrittenQueryStr = "truncate " + getFullTableNameForSQL(tabNameNode);
+    ParseUtils.ReparseResult rr = ParseUtils.parseRewrittenQuery(ctx, rewrittenQueryStr);
+    Context rewrittenCtx = rr.rewrittenCtx;
+    ASTNode rewrittenTree = rr.rewrittenTree;
+
+    BaseSemanticAnalyzer truncate = SemanticAnalyzerFactory.get(queryState, rewrittenTree);
+    // Note: this will overwrite this.ctx with rewrittenCtx
+    rewrittenCtx.setEnableUnparse(false);
+    truncate.analyze(rewrittenTree, rewrittenCtx);
+
+    rootTasks = truncate.getRootTasks();
+    outputs = truncate.getOutputs();
+    updateOutputs(table);
+  }
+
+  private boolean tryMetadataUpdate(Table table, ASTNode tabNameNode, ASTNode whereNode)
+      throws SemanticException {
+    // A feature flag on Hive to perform metadata delete on the source table.
+    if (!HiveConf.getBoolVar(conf, HiveConf.ConfVars.HIVE_OPTIMIZE_METADATA_DELETE)) {
+      return false;
+    }
+    if (whereNode == null || table.getStorageHandler() == null) {
+      return false;
+    }
+    TableName tableName = getQualifiedTableName(tabNameNode);
+    String whereClause = ctx.getTokenRewriteStream().toString(
+        whereNode.getChild(0).getTokenStartIndex(), whereNode.getChild(0).getTokenStopIndex());
+    StringBuilder sb = new StringBuilder("select * from ").append(getFullTableNameForSQL(tabNameNode))
+        .append(" where ").append(whereClause);
+    Context context = new Context(conf);
+    ASTNode rewrittenTree;
+    try {
+      rewrittenTree = ParseUtils.parse(sb.toString(), context);
+    } catch (ParseException pe) {
+      throw new SemanticException(pe);
+    }
+    BaseSemanticAnalyzer sem = SemanticAnalyzerFactory.get(queryState, rewrittenTree);
+    sem.analyze(rewrittenTree, context);
+
+    Map<String, TableScanOperator> topOps = sem.getParseContext().getTopOps();
+    if (!topOps.containsKey(table.getTableName())) {
+      return false;
+    }
+    ExprNodeGenericFuncDesc hiveFilter = topOps.get(table.getTableName()).getConf().getFilterExpr();
+    if (hiveFilter == null) {
+      return false;
+    }
+    ConvertAstToSearchArg.Result result = ConvertAstToSearchArg.createSearchArgument(ctx.getConf(), hiveFilter);
+    if (result.isPartial()) {
+      return false;
+    }
+    SearchArgument sarg = result.getSearchArgument();
+    if (!table.getStorageHandler().canPerformMetadataDelete(table, tableName.getTableMetaRef(), sarg)) {
+      return false;
+    }
```

Comment:
could we extract this to getSarg method that returns an Optional?

Reply:
move sarg construction into a method - Optiional< SearchArgument > getDeleteSarg(....)
````
if(!sarg.IsPresent()) {
    return
}
````

Reply:
Extracted these checks and the Filter predicate conversion to a method. It returns `null` if the the conversion fails. It could return `Optional` but IMHO it is just a matter of coding style. I don's see much difference between a null check and calling `isPresent()`. Optional `sarg` can also be `null` and throw NPE though.

Reply:
👍